test/widgets/lightbox_test.dart

```diff
@@ -197,6 +204,128 @@ class FakeVideoPlayerPlatform extends Fake
 
+  group('LightboxHero', () {
+    late PerAccountStore store;
+    late FakeApiConnection connection;
+
+    final channel = eg.stream();
+    final imageSrcUrlStr = 'https://chat.example/user_uploads/thumbnail/2/ce/nvoNL2LaZOciwGZ-FYagddtK/image.jpg/840x560.webp';
+    final imageSrcUrl = Uri.parse(imageSrcUrlStr);
+    final message = eg.streamMessage(stream: channel,
+      topic: 'test topic', contentMarkdown: ContentExample.imageSingle.html);
+
+    Future<void> setupMessageListPage(WidgetTester tester) async {
+      addTearDown(testBinding.reset);
+      final subscription = eg.subscription(channel);
+      await testBinding.globalStore.add(eg.selfAccount, eg.initialSnapshot(
+        streams: [channel], subscriptions: [subscription]));
+      store = await testBinding.globalStore.perAccount(eg.selfAccount.id);
+      connection = store.connection as FakeApiConnection;
+      await store.addUser(eg.selfUser);
+
+      connection.prepare(json:
+        eg.newestGetMessagesResult(foundOldest: true, messages: [message]).toJson());
+
+      await tester.pumpWidget(TestZulipApp(accountId: eg.selfAccount.id,
+        child: MessageListPage(initNarrow: const CombinedFeedNarrow())));
+
+      await tester.pumpAndSettle();
+    }
+
+    testWidgets('Hero animation occurs smoothly when opening lightbox from message list', (tester) async {
+      prepareBoringImageHttpClient();
+
+      await setupMessageListPage(tester);
+
+      final messageContentFinder = find.byWidgetPredicate((widget) =>
+        widget is MessageContent && widget.message.id == message.id
+      );
+      final messageListImageFinder = find.descendant(
+        of: messageContentFinder,
+        matching: find.byType(RealmContentNetworkImage)
+      );
+      final initialImagePosition = tester.getRect(messageListImageFinder);
+      await tester.tap(messageListImageFinder);
+      await tester.pump();
+      // pump to start hero animation
+      await tester.pump();
+
+      final heroAnimationDuration = Duration(milliseconds: 300);
+      final steps = 150;
+      final stepDuration = heroAnimationDuration ~/ steps;
+      List<Rect> animatedPositions = [];
+      for (int i = 1; i <= steps; i++) {
+        await tester.pump(stepDuration);
+
+        final animatedFlightImageFinder = find.byWidgetPredicate((widget) =>
+          widget is RealmContentNetworkImage && widget.src == imageSrcUrl
+        );
+        animatedPositions.add(tester.getRect(animatedFlightImageFinder));
+      }
+
+      final totalDistance = sqrt(
+        pow(initialImagePosition.top - animatedPositions.last.top, 2) +
+        pow(initialImagePosition.left - animatedPositions.last.left, 2));
+
+      Rect previousPosition = initialImagePosition;
+      double maxStepDistance = 0.0;
+      for (final position in animatedPositions) {
+        final stepDistance = sqrt(
+          pow(position.top - previousPosition.top, 2) +
+          pow(position.left - previousPosition.left, 2));
```

Comment:
The logic here can become a bit easier to see by making a small helper function inside the test case:
```suggestion
        final stepDistance = dist(previousPosition, position);
```

Particularly for comparison with `totalDistance` above.